### DISCUSSION TITLE
Some interesting information regarding Semaphores in C++20 and Wine

### DISCUSSION THREAD

**Discussion by lorn10** [grumpycoders/pcsx-redux#807] [General]
Okay folks (@wheremyfoodat) :smiley:, here follows now the effective explanation while PCSX-Redux is again broken in later Wine versions. 

This has to do with a new functionality in C++20 called **Semaphores** which requires the `msvcp140_atomic_wait.dll` file. At the moment this function is not present in the built-in Wine C++ implementation. So without it, you will get the following known error message:

```
wine: Call from 0000000000272B9E to unimplemented function msvcp140_atomic_wait.dll.__std_atomic_wait_direct, aborting
wine: Unimplemented function msvcp140_atomic_wait.dll.__std_atomic_wait_direct called at address 0000000000272B9E (thread 010c), starting debugger...
wine: Call from 0000000000272B9E to unimplemented function msvcp140_atomic_wait.dll.__std_atomic_notify_one_direct, aborting
```
Interestingly it helped (for a while) to simply put the native `msvcp140_atomic_wait.dll` file manually into the PCSX-Redux program folder. But for whatever reason this stopped to be successful, maybe because my `msvcp140_atomic_wait.dll` became outdated.

Long story short, thanks to CXBX-R developer @PatrickVL it looks to be possible to prevent this problem by avoiding semaphores. I [quote](https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/issues/2347#issuecomment-1099090974):

> Just a thought : this seems to have started with the introduction of `std::binary_semaphore type`, so it could be considered to not use this type and instead use the previous proposal (checking a boolean in a while loop).

I know, because PCSX-Redux is in contrast to CXBX-R available as a native Linux build, it doesn't make sense to change the programming style just to be Wine compliant on the Windows build. Nevertheless its nice to know that "semaphores relationship". :wink: 

Maybe I will bring this to the winetricks GitHub page as a "feature request". So an option to just install the `msvcp140_atomic_wait.dll` file. It is confirmed for Wine 7.0 that it can handle semaphores with the `vcrun2019` parameter, so the installation of native MS Visual C++ 2019. However, this doesn't work anymore for later Wine releases. I can confirm this for Wine devel 7.6.

**Comment by wheremyfoodat** [comment]
Hello @lorn10. I'll probably look into adding running instructions for nice platforms such as Wine and RPi in the near future, since they've got their own slight intricacies. Removing semaphore usage might be a bit of a pain in the ass, so I'd recommend downloading the latest x64 VC++ 2022 redistributable (emphasis on x64) from here https://docs.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170#visual-studio-2015-2017-2019-and-2022 and seeing if it works. Make sure you're using the 64-bit Redux builds as well.

Might want to bring this up to the Wine guys if it doesn't work.

**Reply by lorn10** [reply]
Thanks for your informative answer. :+1: 

I think the corresponding `msvcp140_atomic_wait.dll` will be added to the built-in Wine C++ runtime implementation sometimes in the future. The question is only when this will happen.:wink: So this problem will be effectively resolved when this semaphore C++ 20 feature is implemented in Wine. Will check if that `msvcp140_atomic_wait.dll` problem is already reported at Wines Bugzilla.

And regarding the MS VC++ 2022 redistributable package. These are usually not so easy installable on Wine, the newer the more problematic they are, unfortunately. For example, even with winetricks there exist some strange flaws in conjunction with the MS VC++ 2019 package: https://github.com/Winetricks/winetricks/issues/1736

So based on my experiences with my testing around CXBX-R and PCSX-Redux on newer Wine versions I prefer to not install the native MS VC++ runtime files when ever possible. For me, it really worked there almost perfectly well until someone decided to introduce the `std::binary_semaphore type`. Therefore, as a **non-developer** and just a somehow more technical interested end-user I really dislike this C++ 20 semaphore functionality stuff because it turned a really good working situation (again) into a non-working one. :stuck_out_tongue: :wink: 

But I understand now that this new functionality has absolutely its point and it can help you as a developer quite a lot. And finally, this topic may have also something of "_you have to keep up with the times, otherwise you will go down with them_".

**Update**:  There exist now a bug report at Wine, see [Bug 52821](https://bugs.winehq.org/show_bug.cgi?id=52821).

**Reply by lorn10** [reply]
A final addition to this topic, - I really found out how "Semaphores in C++20" builds are working in newer Wine versions. I quote myself from [here](https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/issues/2347#issuecomment-1100624481):

Second version, **updated on 18.04.2022**

> A short HowTo:
> 
> - simply copy the original `msvcp140_atomic_wait.dll` into the program folder of the affected application.
> 
> In my case, the 32bit `msvcp140_atomic_wait.dll` goes into the `CXBX` folder because CXBX-R is effectively a 32bit application. The 64bit `msvcp140_atomic_wait.dll` comes into the `PCSX-Redux` folder because PCSX-Redux is a 64bit application.
> 
> - start up winecfg from the CLI and select "Libraries", add here a new one called `msvcp140_atomic_wait` (without the dll extension)
> 
> As mentioned in the previous HowTo version, there should be now a new entry "msvcp140_atomic_wait (Native, Builtin)". Exit winecfg and start up CXBX-R build 374ba5e. It should work again. I have tested this also with PCSX-Redux nightly 8457.
> 
> This is also the better solution for the case that the wine-prefix gets updated. And in contrast to the `vcrun2019` winetricks argument this solution is easy reversible (especially for simply "knitted" end-users) and doesn't install the whole native MS Visual C++ 2019 "ballast" when it is effectively not needed. 
> 
> Note, I grabbed the latest msvcp140_atomic_wait.dll file version from my last Windows 7 computer where I have installed the current MSVC runtime library from [here](https://docs.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170). It gave me file version `14.31.31103.0`.
> 
>  PS Due to my [report](https://bugs.winehq.org/show_bug.cgi?id=52821) at Wine Bugzilla, winetricks author Austin English has modified winetricks so that the `msvcp140_atomic_wait` will be installed also on never **Wine devel** versions. A preview can be found [here](https://github.com/austin987/winetricks/commit/9c97d44b9b68a1b09e244e46f9ae0831ab35c315).
